git_ui: Add list of recent commits

jacobtread opened the pull request:
Adds a list of previous commits in the git panel, a simplfiied version of VSCodes "graph" which outlines the commits across branches, this one just shows the ones from the current branch. Very helpful for people like myself who just need to quickly check back across the last couple of commits.

Something similar has been requested a few times already:
- #26511 
- #10807

<img width="581" height="741" alt="image" src="https://github.com/user-attachments/assets/28b01979-f2ae-4c3b-b115-c976e814388c" />

This commit adds the list itself, the setting to enable it, and each item in the list is hooked up to display the commit diff viewer when clicked

Release Notes:

- Added git history list

jacobtread committed:
feat: git history list initial prototype

jacobtread committed:
feat: lazy load in more commits when scrolling

jacobtread committed:
feat: improve position and styling of git history

jacobtread committed:
feat: reload history when repo changes

jacobtread committed:
feat: render scrollbar and reload history on git change

jacobtread committed:
fix: scrolling reset bug

jacobtread committed:
fix: properly handle invalid timestamp errors

jacobtread committed:
feat: setting for commit_history list

jacobtread committed:
refactor: rename git command to align with others

jacobtread committed:
fix: revert unused visibility change

jacobtread commented:
@cla-bot check

jacobtread commented:
@Angelk90 The UI here is intending to mainly replicate the UI from vscode, this is not a full sized panel for viewing history so I'm preferring to keep information here limited to show more commit history in the limited space and less of the less relevant information like commit hashes and dates. As the main use case here is to quickly browse the previous changes and view the diffs for them. 

The information like author, commit hash, and date are all still visible if you click on the item as it opens the diff viewer:
<img width="2052" height="397" alt="image" src="https://github.com/user-attachments/assets/cf7bd2cd-e316-4044-93f4-b226c63b6280" />

jacobtread commented:
> @jacobtread : So if it's like vscode, it loads the first 35 commit messages first, then when you scroll to the end of the 35 commits, the other 35 are loaded?
> How does vscode do it, right?

Yes, in this case I am loading 20 at a time instead of 35 (an arbitrary number I chose, this could be increased or decreased depending on the need but 20 is working fine from my testing)

CharlesChen0823 commented:
I found the prototype for git panel in zed `onboarding` channel, you can reference the [prototype](https://zed-prototyping.vercel.app).

jacobtread committed:
fix: fine tune git reload updates, fix lazy load underflow

Only refresh the git history on repo change events, seems to be a few
other git store events that trigger lots of frequent changes

Fix subtract with overflow

Removes some extra clones

cole-miller reviewed (changes requested):
Hi @jacobtread, thanks for the PR! I think this is looking good, there are just a few blockers to merging it in:

- The "infinite scrolling" doesn't seem to work reliably for me--if I clone https://github.com/pydantic/pydantic-ai-gateway, I can't scroll all the way back to the initial commit.
- Design-wise, we shouldn't render the most recent commit in a different way; it should just be the first entry in the scrolling list (but with the "uncommit" button still present).
- I think this probably shouldn't be behind a setting.

jacobtread committed:
feat: update to use new scrollbar handling

jacobtread committed:
chore: remove unused imports

jacobtread commented:
The infinite scroll works for me, but only when I scroll using a mouse scroll wheel. It appears the scroll bar implementation in Zed does not actually fire the event handler  set by set_scroll_handler when scrolling using the scroll bar (Added logging and the logs only fired when I scrolled using a scroll wheel and not the scrollbar)

jacobtread committed:
feat: remove git history setting show it by default

cole-miller commented:
Thanks, I tried again with your latest revisions and the infinite scroll seems to be working as you describe. If you can fix the issue with the most recent commit being shown twice I think this should be mergeable!

jacobtread commented:
The only problem I have with bringing the current commit into the commit list is how it has a different behavior (more actions and a tooltip) which I can see being out of place in the list as none of the other ones will have this tooltip (Unless the tooltip is revised) since I cannot apply the tooltip to all the list items as in its current state, the tooltip feels quite unintuitive on the list due to the large area around it that keeps it open, making it hard to look through the list as the tooltip gets in the way and you have to move your mouse quite far away to close it

Perhaps instead of combining the two into the list, I leave the clear separation between the two and simply hide the current commit from the list so that it does not appear twice, open to suggestions.

jacobtread committed:
chore: remove old option and extra whitespace

AlexElizard commented:
It would be very good if this PR served as a temporary solution. 

In my opinion, [the git graph](https://github.com/zed-industries/zed/discussions/26511#discussioncomment-13330484) option is preferable.

jacobtread committed:
feat: update to new command system

jacobtread committed:
fix: apply new clippy lints
